### PR TITLE
Make Workspace IDs unique and predictable

### PR DESCRIPTION
**Pull request by hertg**
# Description

Follow up on #848, which made workspace IDs only unique per output, effectively requiring a composite key of `output + id` to identify a workspace. This PR introduces a simple and consistent way of assigning unique IDs to every workspace, as discussed in the comments of #848, particularly in [this comment](https://github.com/leftwm/leftwm/pull/848#issuecomment-1456900361).

**Workspace IDs**
The IDs of workspaces are assigned automatically by LeftWM. They are numeric and counted upwards from $[1..n]$. The workspaces specified in the [config](https://github.com/leftwm/leftwm/wiki/Config#workspaces) receive an ID first. Every configured workspace gets an ID, even if it is on an output that is currently not connected.

```rust
workspaces: [
    ( output: "HDMI-1", y: 0, x: 0, height: 1440, width: 1720 ), // ID: 1
    ( output: "HDMI-1", y: 0, x: 1720, height: 1440, width: 1720 ), // ID: 2
]
```

The count then continues for non-configured / auto-derived workspaces, if there are any. IDs for auto-derived workspaces are assigned in the order that LeftWM receives screen information from the system (xrandr, xinerama, ...).

---

Mergin this PR makes #1027 obsolete

## Type of change

- [x] Development change (no change visible to user)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation only update (no change to the factual codebase)
- [x] This change requires a documentation update

## Updated user documentation:

*TODO* Logic behind automatic workspace ID assignment should probably be documented somewhere

# Checklist:

- [x] Ran `make test-full` locally with no errors or warnings reported
  Note: To fully reproduce CI checks, you will need to run `make test-full-nix`. Usually, this is not neccesary.
- [ ] Manual page has been updated accordingly
- [ ] Wiki pages have been updated accordingly (to perform **after** merge)

### CODE REVIEW

File: leftwm-core/src/handlers/screen_create_handler.rs
Comment (hertg) [inline leftwm-core/src/handlers/screen_create_handler.rs:10]: to be removed

**Comment by hertg** [comment]
Hey @0323pin
Most builds are currently failing because we run everything against `stable` rust channel for linux (which seems to be `1.67.1`). In order to satisfy clippy on this version, a few adjustments have to be made, like using the `let...else` feature. This however makes it so that the msrv action fails, since it is still on `1.60.0`.

So my question is:
1. Can we update the toolchain version for the msrv action, and if so, to what version? Is there a way for us to figure out what the currently latest available rust version is for the OS you are on? (I believe it's OpenBSD?)
2. How can we keep this action up-to-date so we don't run into this issue too much?

**Comment by hertg** [comment]
@0323pin Some additional info

A test by @VuiMuich showed that we'd need to bump the minimum supported rust version to `1.65`. This would allow us to satisfy pedantic clippy on Rust's current `stable` channel. 

If that is not possible we'd need to add some `#[allow(clippy::needless_lifetimes)]` and `#[allow(clippy::manual_let_else)]` to the code.

What is our current philosophy on the minimum supported rust version?

**Comment by 0323pin** [comment]
@hertg thanks for asking.

> Can we update the toolchain version for the msrv action, and if so, to what version?

1.66.1

> Is there a way for us to figure out what the currently latest available rust version is for the OS you are on?

I use NetBSD 😀
But yes, you can find the current available version [here](https://pkgsrc.se/lang/rust) 

> How can we keep this action up-to-date so we don't run into this issue too much?

I'm happy to run test builds, I'm usually running on the latest commit on the main branch but, I'm usually using the latest stable Rust.
So, I'm currenty on 1.67.1 but, the latest version on our repositories is 1.66.1

Generally, NetBSD would be having the latest stable version minus one. Sometimes, it could be two versions behind due to branching. _E.g._ at the end of this month the Q1 stable branch of packages will be released, Rust won't be updated until after the branch.

Hope this helps. Feel free to ask for test builds whenever.

**Comment by hertg** [comment]
@0323pin I just figured that we can have an overview of the latest available rust versions for different OS and distros [on repology](https://repology.org/project/rust/badges).

<details>
<summary>Rust Version Overview</summary>

[Packaging status](https://repology.org/project/rust/versions)
</details>

**Comment by 0323pin** [comment]
Yes, of course. That will do, although as you might have noticed, NetBSD is not listed.

Well, it is but for some reason it's listed as **pkgsrc current**

**Comment by hertg** [comment]
@Eskaan Can you give this a testrun?

File: leftwm-core/src/handlers/screen_create_handler.rs
Comment (Eskaan) [inline leftwm-core/src/handlers/screen_create_handler.rs:10]:
```suggestion
        tracing::trace!("Screen create: {:?}", screen);
```

**Review by Eskaan — CHANGES_REQUESTED** [review]
Along my testing I did not encounter any issues (I did not test as extensively).

Whilst this pr as of itself works, I wanted to leave two suggestions:
1. This pr is breaking for everyone using the current workspace id's (they will change now)
2. There is now a lot of unnecessary checks done that I introduced [(example)](https://github.com/leftwm/leftwm/pull/848/files#diff-a6caf002e502c68d8159046acf3ca4762b238b1b3d3c694a229b914a5ac1eec8L231-R235) where the output gets also matched. These can now be removed as id's are unique.
  I would volunteer to remove them and create a pr on your fork.

**Comment by hertg** [comment]
@Eskaan I fixed the capitalizations across all the logs I found

> I would volunteer to remove them and create a pr on your fork.

Yeah you are very welcome to do that, I think the `output` can also be removed in some other places, like in gutters?

> This pr is breaking [..]

Yes it is, I'll have to also update the CHANGELOG before merging

**Comment by hertg** [comment]
I'm doing this in preparation for #1000 which will likely also be breaking, so those two should be combined within the same release

**Comment by Eskaan** [comment]
https://github.com/hertg/leftwm/pull/3

**Comment by hertg** [comment]
I got the commits from @Eskaan in.
This one is ready for merge IMO.
@VuiMuich any objections?

**Review by VuiMuich — APPROVED** [review]
LGTM

I thought I saw a regression with the config option `single_window_border: false` and `Monocle` Layout, but I guess that is also in the main branch.

**Comment by stevensonmt** [comment]
Was this included in a release yet? I'd really like to get the soft reload bug that puts all open windows on a single tag fixed.

**Comment by Eskaan** [comment]
I don‘t think so. To my knowledge we are waiting for the Layout pr (#1000) to not make multiple releases with breaking changes.

**Comment by stevensonmt** [comment]
When I tried the git version a couple days ago I had the same issue. I'll wait for the update since I don't like having to manually track updates. Especially since I rely on leftwm every day.

**Comment by stevensonmt** [comment]
I went back to the regular version, but I'll look when I get back to that machine and see what was installed.

**Pull request merged**
